actions/run-nextstrain-build-on-aws-batch/action.yaml

```diff
@@ -0,0 +1,68 @@
+name: Run Nextstrain build on AWS Batch
+description: >-
+  This GitHub Actions action is intended to be called by workflows in our other
+  repos when they need to run a nextstrain build on AWS Batch.
+
+inputs:
+  runtime-envvar-names:
+    description: >-
+      A list of environment variable names for environment variables that
+      should be passed to the build. The envvar names should be separated by
+      a space, e.g. "GITHUB_RUN_ID SLACK_TOKEN SLACK_CHANNELS"
+    type: string
+    required: false
+
+  cpus:
+    description: >-
+      Number of CPUs/cores/threads/jobs to utilize at once.
```

Comment:
Similar to my comment about memory, is the recommendation that this be set to one of the [values of a c5 instance](https://aws.amazon.com/ec2/instance-types/) in order to not pay for CPUs which do nothing, or is that not really how instance provisioning works?

Reply:
(Following up your comment from a different thread here, as that thread has moved to a slightly different focus)

> Our compute environments set the [allocation strategy](https://docs.aws.amazon.com/batch/latest/userguide/allocation-strategies.html) to "BEST_FIT" so AWS will try to use an instance that fits the requested the resources at the lowest cost. I've tried to set a job to use 70GiB when the max memory is 72GiB, but AWS still used the next instance up (see [Slack thread](https://bedfordlab.slack.com/archives/C01LCTT7JNN/p1674254476788549)).

> So we just have to estimate the minimum resources needed for a job and AWS will figure out what instance to use.

It would be nice to avoid the caller needing specific knowledge of AWS or our AWS config. I'm thinking here of providing a dropdown of preset memory/cpu combinations which maximise the efficiency of the machines the action uses (currently c5, and the presets would take into account the required headspace noted in that slack thread). If the caller really wants to use their own combination then they can specify them. 

Of course, this could also be done later on as an improvement to the action, it's not necessary from day one. But I would suggest pointing to the AWS c5 docs rather than the nextstrain CLI docs to start with. 

This somewhat couples the action to our AWS config + runner, but I'd think that any advancements there would be accompanied with a new major version release of this action. If the coupling doesn't happen here then it's implicitly happening in the caller actions which I think is worse for maintainability.

Reply:
@joverlee521 [wrote](https://github.com/nextstrain/.github/pull/39#discussion_r1210839259):

> Hmm, maybe we can add the dropdown as part of the workflow-template?

While that might be useful for initial creation of the workflow in a pathogen repo, note that the dropdown won't be available when it comes time to update the CPUs/memory needed. Those subsequent updates are likely to be necessary over time, as they've been several times for existing automated builds.

@jameshadfield [wrote](https://github.com/nextstrain/.github/pull/39#discussion_r1206239089):

> It would be nice to avoid the caller needing specific knowledge of AWS or our AWS config.
> […]
> This somewhat couples the action to our AWS config + runner, but I'd think that any advancements there would be accompanied with a new major version release of this action. If the coupling doesn't happen here then it's implicitly happening in the caller actions which I think is worse for maintainability.

I really don't see how presets vs. direct values reduces coupling for the calling workflow at all. It can't get away from talking about resources one way or another, and as long as those resources are on AWS, it's coupled to AWS.

The value of presets, as I see it, is not reducing coupling but making it easier to choose a resource combination that's most price-efficient for us.

If we wanted to actually reduce coupling, we'd have the calling workflow specify the arbitrary CPUs/memory it wants as a minimum and then have this action fit that into the most price-efficient-for-us CPUs/memory combination.

Reply:
This thread was motivated by me trying to push the PR towards my idea of an interface to running jobs on AWS across many repos. Jover wants this to be basically a wrapper around the Nextstrain CLI, which is fair enough. So these thoughts are no longer PR change requests, just my thoughts. With that out of the way:

> I really don't see how presets vs. direct values reduces coupling for the calling workflow at all. 

I was imagining ~3 parts coming together:
1. Guaranteeing that our current job queue uses c5 instances. If we change this, we'd create a new job queue (and deprecate the old one).
2. Explicitly using that job queue in this workflow, and if we introduce a new queue we'd release a new major version of this workflow (`@v2` etc)
3. This workflow would define resources via a choice of c5 instance, not memory/cpu combos. The job is going to be run on a c5 instance anyway, so why not centralise the mapping of cpu/memory allocations → c5 instance, including our best estimate of the memory overhead needed (as discussed in [this slack thread](https://bedfordlab.slack.com/archives/C01LCTT7JNN/p1674254476788549)). The alternative is that the caller workflows each make a guess of what memory goes with what CPU count (either a total guess, or an educated guess by reading the c5 docs).

> It can't get away from talking about resources one way or another, and as long as those resources are on AWS, it's coupled to AWS.

Yeah, the coupling will always be there, but I'm trying to avoid two situations: (1) where caller workflows make non-ideal choices of memory/cpus, with respect to the available instances and (2) if we change the instances our job queue uses then the caller workflows have no knowledge of that and will be using non-ideal resource combinations. 

> the calling workflow specify the arbitrary CPUs/memory it wants as a minimum and then have this action fit that into the most price-efficient-for-us CPUs/memory combination

That's another (good) option.

Reply:
I see. To me, that's arguing for _tighter_ coupling! Doing that makes sense sometimes, and I see how it would if we're willing to make those guarantees about instance families, job queue versioning, action/workflow versioning, etc.

I'm wary of doing it that way though because those guarantees are extra overhead and though they mean we maintain price-efficiency even after an instance family/job queue switch, we still have to go update all the places (pathogen repos) using the previous versions. I don't think getting absolute price-efficiency is worth that time cost and operational overhead.

I'd rather do the alternative I proposed which _loosens_ coupling and still gets reasonable price-efficiency, because it doesn't require as much ongoing overhead and doesn't require updates to all the callers when we make changes.

> This thread was motivated by me trying to push the PR towards my idea of an interface to running jobs on AWS across many repos. Jover wants this to be basically a wrapper around the Nextstrain CLI, which is fair enough.

I think this will still be useful for running jobs on AWS across many repos…

I agree with Jover's goal here, and see this action/workflow as important "GitHub Actions glue". It reduces the barrier to go from running `nextstrain build` on your computer to running it under automation.